Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift

```diff
@@ -0,0 +1,153 @@
+        }
+
+        let ranges = changedByteRanges(oldTree, rhs: newTree).map { $0.range }
+
+        layer.tree = newTree
+
+        return ranges
+    }
+
+    /// Applies the edit to the current `tree` and returns the old tree and a copy of the current tree with the
+    /// processed edit.
+    /// - Parameters:
+    ///   - tree: The tree before an edit used to parse the new tree.
+    ///   - parser: The parser used to parse the new tree.
+    ///   - edit: The edit to apply.
+    ///   - readBlock: The block to use to read text.
+    /// - Returns: (The old state, the new state).
+    internal func calculateNewState(tree: Tree?,
+                                    parser: Parser,
+                                    edit: InputEdit,
+                                    readBlock: @escaping Parser.ReadBlock) -> (Tree?, Tree?) {
+        guard let oldTree = tree else {
+            return (nil, nil)
+        }
+        semaphore.wait()
+
+        // Apply the edit to the old tree
+        oldTree.edit(edit)
+
+        let newTree = parser.parse(tree: oldTree, readBlock: readBlock)
+
+        semaphore.signal()
+
+        return (oldTree.copy(), newTree)
+    }
+
+    /// Calculates the changed byte ranges between two trees.
+    /// - Parameters:
+    ///   - lhs: The first (older) tree.
+    ///   - rhs: The second (newer) tree.
+    /// - Returns: Any changed ranges.
+    internal func changedByteRanges(_ lhs: Tree?, rhs: Tree?) -> [Range<UInt32>] {
+        switch (lhs, rhs) {
+        case (let t1?, let t2?):
+            return t1.changedRanges(from: t2).map({ $0.bytes })
+        case (nil, let t2?):
+            let range = t2.rootNode?.byteRange
+
+            return range.flatMap({ [$0] }) ?? []
+        case (_, nil):
+            return []
+        }
+    }
+
+    /// Performs an injections query on the given language layer.
+    /// Updates any existing layers with new ranges and adds new layers if needed.
+    /// - Parameters:
+    ///   - textView: The text view to use.
+    ///   - layer: The language layer to perform the query on.
+    ///   - layerSet: The set of layers that exist in the document.
+    ///               Used for efficient lookup of existing `(language, range)` pairs
+    ///   - touchedLayers: The set of layers that existed before updating injected layers.
+    ///                    Will have items removed as they are found.
+    ///   - readBlock: A completion block for reading from text storage efficiently.
+    /// - Returns: An index set of any updated indexes.
+    @discardableResult
+    internal func updateInjectedLanguageLayers(textView: HighlighterTextView,
+                                               layer: LanguageLayer,
+                                               layerSet: inout Set<LanguageLayer>,
+                                               touchedLayers: inout Set<LanguageLayer>,
+                                               readBlock: @escaping Parser.ReadBlock) -> IndexSet {
```

Comment:
### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
    internal func updateInjectedLanguageLayers(
        textView: HighlighterTextView,
        layer: LanguageLayer,
        layerSet: inout Set<LanguageLayer>,
        touchedLayers: inout Set<LanguageLayer>,
        readBlock: @escaping Parser.ReadBlock
    ) -> IndexSet {
```